File: src/services/audit/detectors/ad/advanced/monitoring/audit-policy-weak.ts
/**
 * Audit Policy Weak Detector
 * Check if audit policy is weak/incomplete
 * Requires GPO settings from SYSVOL
 */

import { ADDomain } from '../../../../../../types/ad.types';
import { Finding } from '../../../../../../types/finding.types';
import { GpoSecuritySettings } from '../types';

export function detectAuditPolicyWeak(
  gpoSettings: GpoSecuritySettings | null,
  domain: ADDomain | null,
  includeDetails: boolean
): Finding {
  // Critical audit categories that should be enabled
  const criticalAuditCategories = [
    'Account Logon',
    'Account Management',
    'Logon/Logoff',
    'Object Access',
    'Policy Change',
    'Privilege Use',
    'System',
  ];

  if (gpoSettings && gpoSettings.auditPolicies && gpoSettings.auditPolicies.length > 0) {
    const configuredCategories = new Set(gpoSettings.auditPolicies.map((p) => p.category));
    const missingCategories = criticalAuditCategories.filter((cat) => !configuredCategories.has(cat));

    // Check if critical events are being audited
    const hasWeakAudit = missingCategories.length > 0;

    return {
      type: 'AUDIT_POLICY_WEAK',
      severity: 'medium',
      category: 'advanced',
      title: 'Audit Policy Incomplete',
      description:
        'Domain audit policy does not cover all critical security events. Attacks may go undetected.',
      count: hasWeakAudit ? 1 : 0,
      affectedEntities: includeDetails && hasWeakAudit && domain ? [domain.dn] : undefined,
      details: hasWeakAudit
        ? {
            recommendation: 'Configure Advanced Audit Policy to audit all critical security categories.',
            missingCategories,
            configuredCategories: Array.from(configuredCategories),
          }
        : undefined,
    };
  }

  return {
    type: 'AUDIT_POLICY_WEAK',
    severity: 'medium',
    category: 'advanced',
    title: 'Audit Policy Configuration Unknown',
    description: 'Unable to determine audit policy configuration. Manual review recommended.',
    count: 0,
    details: {
      note: 'GPO audit settings not available. Check Advanced Audit Policy Configuration manually.',
      requiredCategories: criticalAuditCategories,
    },
  };
}

File: internal/audit/detectors/ad/advanced/monitoring/audit-policy-weak.go
package monitoring

import (
	"context"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// AuditPolicyWeakDetector detects weak or incomplete audit policies
type AuditPolicyWeakDetector struct {
	audit.BaseDetector
}

// NewAuditPolicyWeakDetector creates a new detector
func NewAuditPolicyWeakDetector() *AuditPolicyWeakDetector {
	return &AuditPolicyWeakDetector{
		BaseDetector: audit.NewBaseDetector("AUDIT_POLICY_WEAK", audit.CategoryAdvanced),
	}
}

// Detect executes the detection
func (d *AuditPolicyWeakDetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	// Critical audit categories that should be enabled
	criticalAuditCategories := []string{
		"Account Logon",
		"Account Management",
		"Logon/Logoff",
		"Object Access",
		"Policy Change",
		"Privilege Use",
		"System",
	}

	// This would require GPO settings from SYSVOL - for now, flag for manual review
	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityMedium,
		Category:    string(d.Category()),
		Title:       "Audit Policy Configuration Unknown",
		Description: "Unable to determine audit policy configuration. Manual review recommended.",
		Count:       0,
		Details: map[string]interface{}{
			"note":               "GPO audit settings not available via LDAP. Check Advanced Audit Policy Configuration manually.",
			"requiredCategories": criticalAuditCategories,
		},
	}

	return []types.Finding{finding}
}

func init() {
	audit.MustRegister(NewAuditPolicyWeakDetector())
}
